format/bplist/bplist.go

```diff
@@ -0,0 +1,245 @@
+package bplist
+
+import (
+	"math"
+	"time"
+
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/interp"
+	"github.com/wader/fq/pkg/scalar"
+)
+
+func init() {
+	interp.RegisterFormat(decode.Format{
+		Name:        format.BPLIST,
+		ProbeOrder:  format.ProbeOrderBinFuzzy, // after most others (overlap some with webp)
+		Description: "Apple Binary Property List",
+		Groups:      []string{format.PROBE},
+		DecodeFn:    bplistDecode,
+	})
+}
+
+const (
+	elementTypeNullOrBoolOrFill = 0x00
+	elementTypeInt              = 0x01
+	elementTypeReal             = 0x02
+	elementTypeDate             = 0x03
+	elementTypeData             = 0x04
+	elementTypeASCIIString      = 0x05
+	elementTypeUnicodeString    = 0x06
+	elementTypeUID              = 0x08
+	elementTypeArray            = 0x0A
+	elementTypeSet              = 0x0C
+	elementTypeDict             = 0x0D
+)
+
+const (
+	null      = 0x00
+	boolFalse = 0x08
+	boolTrue  = 0x09
+)
+
+var singleByteElementMap = scalar.UToScalar{
+	null:      {Sym: "null", Description: "Null Value"},
+	boolFalse: {Sym: "false", Description: "False Value"},
+	boolTrue:  {Sym: "true", Description: "True Value"},
+}
+
+var elementTypeMap = scalar.UToScalar{
+	elementTypeNullOrBoolOrFill: {Sym: "singleton", Description: "Singleton value (null/bool)"},
+	elementTypeInt:              {Sym: "int", Description: "Integer"},
+	elementTypeReal:             {Sym: "real", Description: "Floating Point Number"},
+	elementTypeDate:             {Sym: "date", Description: "Date, 8-byte float"},
+	elementTypeData:             {Sym: "data", Description: "Binary data"},
+	elementTypeASCIIString:      {Sym: "ascii_string", Description: "ASCII encoded string"},
+	elementTypeUnicodeString:    {Sym: "unicode_string", Description: "Unicode string"},
+	elementTypeUID:              {Sym: "uid", Description: "UID"},
+	elementTypeArray:            {Sym: "array", Description: "Array"},
+	elementTypeSet:              {Sym: "set", Description: "Set"},
+	elementTypeDict:             {Sym: "dict", Description: "Dictionary"},
+}
+
+// decodes the number of bits required to store the following object
+func decodeSize(d *decode.D) uint64 {
```

Comment:
If separate parts of the size encoding is not important one can decode have own scalar decoders. Would look like this functions but skip Field* calls and then use like this `d.UFn("size", decodeSize)`

Reply:
For purely exploratory purposes and full transparency, a user might want to see the distinct elements of that size encoding. On the other hand it does create another level of nesting. I'll defer to your judgment on which way to go since you know what has been done elsewhere.

Reply:
Yes it's a bit hard know where to draw the line sometimes. But if this integer encoding is very specific to bplist i think it might make sense to decode it in parts? sometimes i've also added a "synthetic" field (has no bit range) to the struct with the value it encodes, ex see the asn1 ber decoders use of `d.FieldValue*`, that way the value is also accessible in queries.

One thing i've thought about it to add some kind of metadata per scalar value so that you can inspect what kind of encoding it uses.

Reply:
I'll add a synthetic `size` field for each entry to make it easier to query, and rename the `size` field in `decodeSize` to `size_bigint`.